source/common/config/utility.h

```diff
@@ -275,45 +277,32 @@ class Utility {
   }
 
   /**
-   * Translate a nested config into a route-specific proto message provided by
-   * the implementation factory.
-   * @param source Protobuf::Message containing the opaque config for the given factory's
-   *        route-local configuration.
-   * @param factory Server::Configuration::NamedHttpFilterConfigFactory implementation
-   * @return ProtobufTypes::MessagePtr the translated config
+   * Translate an opaque config map into a map of config objects provided by implementation factory.
+   * @param opaque_config_map the proto map from name to opaque config struct.
+   * @param factory_lookup a function to lookup factory from name.
+   * @param message_factory a function to create empty proto message from the factory.
+   * @param object_factory a function to create the config object from the message.
    */
-  static ProtobufTypes::MessagePtr
-  translateToFactoryRouteConfig(const Protobuf::Message& source,
-                                Server::Configuration::NamedHttpFilterConfigFactory& factory) {
-    ProtobufTypes::MessagePtr config = factory.createEmptyRouteConfigProto();
-
-    // Fail in an obvious way if a plugin does not return a proto.
-    RELEASE_ASSERT(config != nullptr, "");
-
-    MessageUtil::jsonConvert(source, *config);
-    return config;
-  }
-
-  /**
-   * Translate a nested config into a protocol-specific options proto message provided by the
-   * implementation factory.
-   * @param source Protobuf::Message containing the opaque config for the given factory's
-   *        protocol specific configuration.
-   * @param factory Server::Configuration::NamedNetworkFilterConfigFactory implementation
-   * @return ProtobufTypes::MessagePtr the translated config
-   * @throws EnvoyException if the factory does not support protocol options
-   */
-  static ProtobufTypes::MessagePtr
-  translateToFactoryProtocolOptionsConfig(const Protobuf::Message& source, const std::string& name,
-                                          Server::Configuration::ProtocolOptionsFactory& factory) {
-    ProtobufTypes::MessagePtr config = factory.createEmptyProtocolOptionsProto();
-
-    if (config == nullptr) {
-      throw EnvoyException(fmt::format("filter {} does not support protocol options", name));
+  template <class ConfigObject, class Factory>
+  static absl::flat_hash_map<std::string, ConfigObject> translateOpaqueConfigMap(
```

Comment:
I think this callback oriented approach is kind of hard to follow. Could we have a virtual interface at the expense of it being a bit more verbose? Otherwise, this starts to look like Javascript :)

Reply:
Yeah I know this is not very clean, but a virtual interface will be templated and introduce much more boiler plate so it might not worth. Another option is do those as static functions for `ConfigObject` and use template. Thought it is hard to follow anyway. WDYT?

I'm fine to revert the refactor part and just introduce flat_hash_map too.

Reply:
Introducing flat_hash_map separately seems like a win.

BTW I know the general recommendation is to use flat_hash_map for string, but I didn't find that to be optimal. It might be good to try node_hash_map and see which is faster/smaller.

You can't really go too wrong with flat_hash_map, but the usual motivation to switch is to improve performance it might be good to measure it.

OTOH if this isn't performance sensitive at all and you are just doing it to be consistent then carry on :)

Reply:
@jmarantz Per https://abseil.io/blog/20180927-swisstables

> We generally recommend that you use absl::flat_hash_map<K, std::unique_ptr<V>> instead of absl::node_hash_map<K, V>.

So I think we could go with flat_hash_map as the ConfigObject is wrapped in const shared ptr.

The write part (this part) isn't performance critical, the getter part is on the data path.

Reply:
Yup should work fine with that value. I just found it was slower/bigger than node_hash_map due to the complexity of the std;:string as key. Works fine though, and if perf is not a consideration then this is fine.

Reply:
@lizan I think just doing some virtual class hierarchy with the concrete types and then using them in the template concretely with the instantiated types? I agree that in general, virtual inheritance and templates don't work well.

I'm in the Anything But Callbacks Everywhere camp :)